### PR TITLE
Move nn.module state dict pre hook 

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98964

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
This would need test!

cc @rohan-varma

**Comment by zhaojuanmao** [comment]
added a unit test for lazy module, without this change, lazy module unit test will fail

File: test/test_nn.py [hunk 1]
```diff
@@ -2425,10 +2425,35 @@ def my_state_dict_pre_hook(*args, **kwargs):
...
         self.assertTrue(called)
 
-    def test_register_state_dict_pre_hook(self):
+    def _test_register_state_dict_pre_hook(self, model, submodule):
         _state_dict_prefix = "foo."
         state_dict_pre_hook_count = 0
+        keep_var_setting = False
 
+        def my_state_dict_pre_hook(module, prefix, keep_vars):
+            nonlocal keep_var_setting
```
Comment (albanD) [inline test/test_nn.py:2434]: This is not needed if you only read the variable.

File: test/test_nn.py [hunk 2]
```diff
@@ -2425,10 +2425,35 @@ def my_state_dict_pre_hook(*args, **kwargs):
...
+            nonlocal state_dict_pre_hook_count
+            state_dict_pre_hook_count += 1
+            self.assertTrue(prefix.startswith(_state_dict_prefix))
+
+        model.register_state_dict_pre_hook(my_state_dict_pre_hook)
+        # Test to ensure submodules run the hook as well.
+        submodule.register_state_dict_pre_hook(my_state_dict_pre_hook)
+
+        def check_results(model):
+            nonlocal _state_dict_prefix, state_dict_pre_hook_count, keep_var_setting
```
Comment (albanD) [inline test/test_nn.py:2445]: Similar only needed for the things you actually modify

**Review by albanD — COMMENTED** [review]
Only small nits on testing, sounds good otherwise.

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**